{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# epann: Evolving Plastic Neural Networks\n",
    "\n",
    "\n",
    "Elan,\n",
    "\n",
    "Here is a step by step of how to work with some aspects of the package to instantiate CPPN genomes in response to your email. \n",
    "\n",
    "Let me know if anything remains unclear.\n",
    "\n",
    "Chad\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Definition: \n",
    "\n",
    "**Compositional Pattern Producing Networks**\n",
    "\n",
    "***CPPN***s are an *abstraction of inheritance and development* constructed from simple artificial neural networks. \n",
    "\n",
    "Their goal is take in a possible connection in a final agent's brain as coordinates within a defined ***substrate***, and output some feature of that connection, such as it's weight and learning parameters. \n",
    "\n",
    "By representing an agent this way, it is not necessary to encode every connection in the final agent phenotype when performing evolutionary computation to find good solutions. \n",
    "\n",
    "Instead, we can *indirectly encode* agents with much smaller genotypes that describe the simple CPPN genome. An indirect encoding simplifies our representation of a solution, and allows us to search through a larger portion of the space of possible solutions to solve a task.\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example\n",
    "\n",
    "Let's start with a simple example using the **epann** package. \n",
    "\n",
    "We will construct a population of 5 agents with very simple genome neural networks (CPPNs) to demonstrate how they are initialized, how their structure is related to the final agent brain, and how they change (mutation and reproduction) over the course of evolution. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Agent 0 - <epann.core.population.genome.cppn.CPPN instance at 0x7f8237d78e60>\n",
      "Agent 1 - <epann.core.population.genome.cppn.CPPN instance at 0x7f823dfde998>\n",
      "Agent 2 - <epann.core.population.genome.cppn.CPPN instance at 0x7f8235839d88>\n",
      "Agent 3 - <epann.core.population.genome.cppn.CPPN instance at 0x7f823e06f6c8>\n",
      "Agent 4 - <epann.core.population.genome.cppn.CPPN instance at 0x7f823df51cf8>\n"
     ]
    }
   ],
   "source": [
    "from epann.core.population.population import Population\n",
    "\n",
    "num_agents = 5\n",
    "\n",
    "pop = Population(num_agents)\n",
    "\n",
    "for agent in pop.genomes.keys():\n",
    "    print 'Agent', agent, '-', pop.genomes[agent]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As you can see, each agent is defined as an instance of a CPPN object. Within this object are attributes that define its genotype, which can then be used to construct a phenotype for the agent. \n",
    "\n",
    "Let's set aside the first agent (**Agent 0**) and take a look at this genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "current_agent = pop.genomes[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Most importantly for our discussion, the current agent has two sets of genome lists that can be modified over the course of evolution among its attributes. \n",
    "\n",
    "It has a *node genome*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{0: {'activation': 'linear', 'type': 'input'}, 1: {'activation': 'linear', 'type': 'input'}, 2: {'activation': 'linear', 'type': 'input'}, 3: {'activation': 'linear', 'type': 'input'}, 4: {'activation': 'linear', 'type': 'input'}, 5: {'activation': 'ramp', 'type': 'output'}}\n"
     ]
    }
   ],
   "source": [
    "print current_agent.nodes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It's *node genome* is a dictionary of genes that describe the characteristics of individual nodes in the CPPN.\n",
    "\n",
    "For example, **Node 5** is an output node ('type') with a unique activation function ('activation'). \n",
    "\n",
    "(**Note:** it might seem odd that an output node does not have a more traditional activation function, such as the *sigmoid*. Neurons in CPPNs can have a variety of activation functions that are selected for their ability to introduce repetition or symmetry, which gives rise to the network's *pattern producing* capabilities. More on this distinction later.) \n",
    "\n",
    "For now, we can at least observe the possible activation functions output nodes can be assigned to:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['x_cubed', 'linear', 'sigmoid', 'ramp', 'gauss', 'abs_value', 'tan_h', 'step', 'ReLU', 'sine']\n"
     ]
    }
   ],
   "source": [
    "from epann.core.tools.utils.activations import Activation\n",
    "acts = Activation()\n",
    "print acts.tags"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nodes within the CPPN (except for the input nodes) can have any of these activation functions. For now, let's set the activation function to something simple. This will become clear why when we get to explaining the substrate, and we will change it back when we're done. Once you feel like you have gotten the hang of the relationship between the substrate and the CPPN genome, you can set the output node activation to any of the strings in the above list and see what the substrates look like (then just select Run All from the Cell pull down menu). \n",
    "\n",
    "Note: some values are not set up to be sampled with more than one input value (i.e. a meshgrid), so just play around with it a bit to see which ones are working properly. (for example, use 'step', not 'ReLU')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Save the old randomly generated activation function\n",
    "old_output_act = current_agent.nodes[5]['activation']\n",
    "\n",
    "# Re-assign the output node activation function\n",
    "current_agent.nodes[5]['activation'] = 'step'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We start a generation off with 5 agents that have the same number of input and output nodes. As a result, every agent will have identical *node genomes* when they are intialized, save the specific activation functions assigned to the output nodes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Input nodes are equivalent across the population when initialized...\n",
      "\n",
      "- Agent 0\n",
      "    Node 0 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 1 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 2 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 3 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 4 : {'activation': 'linear', 'type': 'input'}\n",
      "- Agent 1\n",
      "    Node 0 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 1 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 2 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 3 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 4 : {'activation': 'linear', 'type': 'input'}\n",
      "- Agent 2\n",
      "    Node 0 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 1 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 2 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 3 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 4 : {'activation': 'linear', 'type': 'input'}\n",
      "- Agent 3\n",
      "    Node 0 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 1 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 2 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 3 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 4 : {'activation': 'linear', 'type': 'input'}\n",
      "- Agent 4\n",
      "    Node 0 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 1 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 2 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 3 : {'activation': 'linear', 'type': 'input'}\n",
      "    Node 4 : {'activation': 'linear', 'type': 'input'}\n",
      "\n",
      "While output nodes differ in their specific activation functions...\n",
      "\n",
      "- Agent 0\n",
      "    Node 5 : {'activation': 'step', 'type': 'output'}\n",
      "- Agent 1\n",
      "    Node 5 : {'activation': 'sine', 'type': 'output'}\n",
      "- Agent 2\n",
      "    Node 5 : {'activation': 'sigmoid', 'type': 'output'}\n",
      "- Agent 3\n",
      "    Node 5 : {'activation': 'ramp', 'type': 'output'}\n",
      "- Agent 4\n",
      "    Node 5 : {'activation': 'x_cubed', 'type': 'output'}\n"
     ]
    }
   ],
   "source": [
    "# Input nodes\n",
    "print '\\nInput nodes are equivalent across the population when initialized...\\n'\n",
    "for agent in range(num_agents):\n",
    "    print '- Agent', agent\n",
    "    for node in range(5):\n",
    "        print '    Node', node, ':', pop.genomes[agent].nodes[node]\n",
    "        \n",
    "# Output nodes\n",
    "print '\\nWhile output nodes differ in their specific activation functions...\\n'\n",
    "for agent in range(num_agents):\n",
    "    print '- Agent', agent\n",
    "    print '    Node', 5, ':', pop.genomes[agent].nodes[5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The CPPN also has a *connection genome* that keeps track of the connections between these nodes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{0: {'enable_bit': 1, 'in_node': 5, 'weight': -0.6694964313532505, 'out_node': 0}, 1: {'enable_bit': 1, 'in_node': 5, 'weight': 1.82539028004255, 'out_node': 1}, 2: {'enable_bit': 1, 'in_node': 5, 'weight': 0.8763862342734189, 'out_node': 2}, 3: {'enable_bit': 1, 'in_node': 5, 'weight': 1.2931369895594522, 'out_node': 3}, 4: {'enable_bit': 1, 'in_node': 5, 'weight': 0.802118384963248, 'out_node': 4}}\n"
     ]
    }
   ],
   "source": [
    "print current_agent.connections"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just like any neural network you are accustomed to seeing, a CPPN is composed of an input layer (**Nodes 0-4**) and an output layer (with a single output node, **Node 5**). \n",
    "\n",
    "Each agent begins with 6 total nodes which are *fully connected*, making 5 initial weights. Although agents in the population are *structurally identical* (they have the same number of initial nodes in their CPPN), the weights of these connections will not be the same for each agent.\n",
    "\n",
    "Let's compare a single connection across the population - **Connection 0**, between **Node 0** and **Node 5** to show this fact:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Connection weights are randomly initialized across the population for the same connection...\n",
      "\n",
      "- Agent 0\n",
      "    Connection 0 : {'enable_bit': 1, 'in_node': 5, 'weight': -0.6694964313532505, 'out_node': 0}\n",
      "- Agent 1\n",
      "    Connection 0 : {'enable_bit': 1, 'in_node': 5, 'weight': 0.014733169930874333, 'out_node': 0}\n",
      "- Agent 2\n",
      "    Connection 0 : {'enable_bit': 1, 'in_node': 5, 'weight': 1.1183923189249478, 'out_node': 0}\n",
      "- Agent 3\n",
      "    Connection 0 : {'enable_bit': 1, 'in_node': 5, 'weight': -0.05951662805764614, 'out_node': 0}\n",
      "- Agent 4\n",
      "    Connection 0 : {'enable_bit': 1, 'in_node': 5, 'weight': -1.2581991745370176, 'out_node': 0}\n"
     ]
    }
   ],
   "source": [
    "print 'Connection weights are randomly initialized across the population for the same connection...\\n'\n",
    "for agent in range(num_agents):\n",
    "    print '- Agent', agent\n",
    "    print '    Connection', 0, ':', pop.genomes[agent].connections[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, the connection genome is really where the meat of the Hyper-NEAT algorithm lies, so let's focus on it a little more. \n",
    "\n",
    "Here is what this CPPN genome looks like graphically:\n",
    "\n",
    "### Figure 1. Initialized CPPN genome\n",
    "\n",
    "![Figure 1](epann/figures/init_cppn.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each of the nodes have an identification number specific to its node genome (node IDs are written in white). \n",
    "\n",
    "Connections also have identification numbers, which we will refer to as *innovation numbers* (innovation numbers written along connections in black). \n",
    "\n",
    "This example CPPN is an agent's *genome*. Since neural networks act as function-approximators, an agent's genome is in effect, also a function. In this case, the genome is a function that maps connections in the agent's final brain to the strength of those connections (the weight). \n",
    "\n",
    "I will go into some more detail below describing how this is done, but for now it is important to remember that the genome of an agent is a function represented as an artificial neural network. That genome neural network will be used to construct the agent's brain, which is also a neural network. \n",
    "\n",
    "In an attempt to ensure that these two neural networks are not confused, CPPN nodes (the genotype) are drawn in blue. CPPN nodes have identification numbers written in white, and connections between nodes have innovation numbers written in black. Agent neural networks (their \"brains\"), do not have identification numbers for their nodes or connections (they have coordinates, described below). Furthermore, their nodes are drawn in red or orange.\n",
    "\n",
    "Let's look at the first connection (**Connection 0**; innovation number = 0) in this **Agent 0**'s connection genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 - {'enable_bit': 1, 'in_node': 5, 'weight': -0.6694964313532505, 'out_node': 0}\n"
     ]
    }
   ],
   "source": [
    "print 0, '-', current_agent.connections[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we can verify in **Figure 1** (innovation # label 0), **Connection 0** describes a connection between **Node 0** and **Node 5**. \n",
    "\n",
    "These landing points are intuitively labeled within the gene above: **Connection 0** goes *out* from **Node 0**, so it is labelled the 'out_node' of that connection. **Connection 0** terminates *into* **Node 5**, and is thereby **Connection 0**'s 'in_node'. \n",
    "\n",
    "You can go over the connection genome for **Agent 0** to verify that the rest of the connections share this convention.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(innovation # / connection#) : 0 - {'enable_bit': 1, 'in_node': 5, 'weight': -0.6694964313532505, 'out_node': 0}\n",
      "(innovation # / connection#) : 1 - {'enable_bit': 1, 'in_node': 5, 'weight': 1.82539028004255, 'out_node': 1}\n",
      "(innovation # / connection#) : 2 - {'enable_bit': 1, 'in_node': 5, 'weight': 0.8763862342734189, 'out_node': 2}\n",
      "(innovation # / connection#) : 3 - {'enable_bit': 1, 'in_node': 5, 'weight': 1.2931369895594522, 'out_node': 3}\n",
      "(innovation # / connection#) : 4 - {'enable_bit': 1, 'in_node': 5, 'weight': 0.802118384963248, 'out_node': 4}\n"
     ]
    }
   ],
   "source": [
    "for connection in current_agent.connections.keys():\n",
    "    print '(innovation # / connection#) :', connection, '-', current_agent.connections[connection]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each connection has a weight labeled by the key 'weight'. This value is randomly generated for each agent and connection in every experiment. \n",
    "\n",
    "A label which may seem out of place is the 'enable_bit'. In our genetic algorithm (Hyper-NEAT), mutations will cause the topology of this CPPN to change over time. **Connection 0** (which goes from **Node 0** to **Node 5**) could be replaced by other connections that connect it to different nodes. If this occures, then the connection will no longer be expressed. \n",
    "\n",
    "It will become clear once we get deeper into the mutations that can happen to a genome how labelling connections with *innovation numbers* becomes useful in experiments, but basically these labels allow us to compare the evolutionary histories of a particular agent during the reproduction stages of our genetic algorithm. \n",
    "\n",
    "We can line up innovation numbers between two parents, and have some confidence that they share similar histories of mutations if their innovation numbers are similar. \n",
    "\n",
    "Because we would like to be able to compare agents to one another this way, it would not be advantageous to, for example, delete a mutation entirely from a genome. Instead, we can *flip the enable_bit* gene in a genome, that is reassign the 'enable_bit' attribute for a connection to 0. Doing this turns off this connection from being expressed in the genome without deleting it entirely. \n",
    "\n",
    "Perhaps this is a lot to cover without seeing a more concrete example, so lets take a moment to break a part the relationship between a CPPN genome and the final agent neural network phenotype that we want to construct and evaluate on a task."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The Substrate\n",
    "\n",
    "An important starting point when thinking about our CPPN abstraction is that we are going to treat our final agent neural network, the brain that will actually be interacting with and learning from the task we are interested in, as something that exists in a two-dimensional space. As a consequence, individual nodes in that network can be defined with locations within that space. \n",
    "\n",
    "### A simple task\n",
    "\n",
    "A very simple task that we might be interested in having our agents solve is one modified off the game Flappy Bird. (a single frame from this task is shown below in **Figure 2**)\n",
    "\n",
    "An agent remains stationary in a column of a grid world, while the world continues to fly by it, updating the values of columns with each time step. The agent in **Figure 2** below is black, nutrients in the world are green, and blue pixels indicate the agent's field of view (what the organism can actually see around it).\n",
    "\n",
    "While the agent always remains in the same column, it can execute one of three actions at any time based on what it sees in the environment to change its row. It can stay in its current row, or it can move upwards or downwards one row. \n",
    "\n",
    "We would like to find agents who can feed themselves - that is, organisms that can observe nutrient locations and execute the appropriate actions to pass through those pixels and ingest them. (One potential objective is to find agent's who eat the most nutrients in a given trial)\n",
    "\n",
    "\n",
    "If we designed a neural network for this controller, it would require at least six input nodes (one for each of the pixels in its field of view) and three output nodes (one for each of the potential actions). \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 2. Example Control task\n",
    "\n",
    "\n",
    "![Figure 1](epann/figures/agent.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The substrate\n",
    "\n",
    "Now that we have chosen a simple task for our agents, let's return to defining a ***substrate***, which defines nodes in our agent's brain as having locations in a two-dimensional space. Let's take a moment and visualize the nodes we know our agent must contain (6 input nodes and 3 output nodes) in this space:\n",
    "\n",
    "### Figure 3. Substrate Coordinate Space\n",
    "\n",
    "![Figure 3](epann/figures/substrate1.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remember, this space and the neural network it contains is different than the Compositional Pattern Producing Network (CPPN) genome shown before. CPPNs do not live in this space - it is a space where we can assign locations to nodes in our agent's \"brain\". Nodes in this network are drawn in red to further distinguish them from the CPPN (which were drawn in blue). Like any other controller neural network, an observation from the environment is used as these input nodes, and it outputs a selected action.\n",
    "\n",
    "In **Figure 3**, we can see our 6 input nodes are aligned across the bottom and our 3 output nodes are likewise aligned across the top. \n",
    "\n",
    "The space itself is continuous between -1 and 1 in both $x$ and $y$. ( $ x, y \\in [-1, 1] $ )\n",
    "\n",
    "From this perspective, *input nodes* can be defined with coordinates that describe their *location* along the bottom edge of this space: $(x, -1)$. If we assume that these nodes are evenly spaced along the edge, we can initialize their locations as:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 360,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[-1.0, -1.0], [-0.59999999999999998, -1.0], [-0.19999999999999996, -1.0], [0.20000000000000018, -1.0], [0.60000000000000009, -1.0], [1.0, -1.0]]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "input_locations = [ [x, -1.0] for x in np.linspace(-1, 1, 6) ]\n",
    "print input_locations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Similarly, the three output nodes exist along the top edge of the substrate: $(x, +1)$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 361,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[-1.0, 1.0], [0.0, 1.0], [1.0, 1.0]]\n"
     ]
    }
   ],
   "source": [
    "output_locations = [ [ x, 1.0 ] for x in np.linspace(-1, 1, 3) ]\n",
    "print output_locations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, the purpose of our CPPN genome is to allow us to more compactly represent an agent's 'brain' as an indirect encoding. The way we have chosen to implement this mapping from genotype to phenotype involves sampling a possible connection between two nodes in the *agent's brain* and feeding it through the CPPN genome function as input. In return, we receive the weight between those two nodes. \n",
    "\n",
    "Placing nodes into this substrate space allows us to represent individual nodes in this connection with coordinates, as described above. So let's try a connection given our current initialized CPPN for **Agent 0**. (See visualization on the substrate in **Figure 4**)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### Figure 4: A sample connection on the substrate between an input and output node\n",
    "\n",
    "![Figure 4](epann/figures/substrate1_connect0.png)\n",
    "\n",
    "The first input node is located at $[-1.0, -1.0]$, and we are going to determine the weight between it and the first output node, which is located at $[-1.0, 1.0]$ by feeding this connection into our CPPN genome function. \n",
    "\n",
    "If we look again at **Figure 1**, we can see that we replace our inputs to the CPPN:\n",
    "\n",
    "$$ (x_1, y_1, x_2, y_2) = (-1.0, -1.0, -1.0, 1.0) $$\n",
    "\n",
    "If we evaluate our initialized CPPN genome for this connection, we receive the weight of this connection as output.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 362,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "- From input node [-1.0, -1.0] to output node [-1.0, 1.0] :\n",
      "    * weight =  1.0\n"
     ]
    }
   ],
   "source": [
    "from cppn_overview import InitializedAgent\n",
    "\n",
    "agent = InitializedAgent(current_agent)\n",
    "\n",
    "# Define the current connection\n",
    "connection_start = input_locations[0]\n",
    "connection_end = output_locations[0]\n",
    "\n",
    "# Evaluate the connection\n",
    "print '- From input node', connection_start, 'to output node', connection_end, ':'\n",
    "print '    * weight = ', agent.evaluate(connection_start, connection_end)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we revisit **Figure 1**, we can see that this output value describes the weight of the connection we are sampling between the input node and the output node. \n",
    "\n",
    "We can sample again across all of our input nodes to evaluate the weights to each output node."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "From input node [-1.0, -1.0]\n",
      "   - to output node [-1.0, 1.0] :\n",
      "      * weight =  1.0\n",
      "   - to output node [0.0, 1.0] :\n",
      "      * weight =  1.0\n",
      "   - to output node [1.0, 1.0] :\n",
      "      * weight =  1.0\n",
      "\n",
      "From input node [-0.59999999999999998, -1.0]\n",
      "   - to output node [-1.0, 1.0] :\n",
      "      * weight =  -1.0\n",
      "   - to output node [0.0, 1.0] :\n",
      "      * weight =  1.0\n",
      "   - to output node [1.0, 1.0] :\n",
      "      * weight =  1.0\n",
      "\n",
      "From input node [-0.19999999999999996, -1.0]\n",
      "   - to output node [-1.0, 1.0] :\n",
      "      * weight =  -1.0\n",
      "   - to output node [0.0, 1.0] :\n",
      "      * weight =  1.0\n",
      "   - to output node [1.0, 1.0] :\n",
      "      * weight =  1.0\n",
      "\n",
      "From input node [0.20000000000000018, -1.0]\n",
      "   - to output node [-1.0, 1.0] :\n",
      "      * weight =  -1.0\n",
      "   - to output node [0.0, 1.0] :\n",
      "      * weight =  1.0\n",
      "   - to output node [1.0, 1.0] :\n",
      "      * weight =  1.0\n",
      "\n",
      "From input node [0.60000000000000009, -1.0]\n",
      "   - to output node [-1.0, 1.0] :\n",
      "      * weight =  -1.0\n",
      "   - to output node [0.0, 1.0] :\n",
      "      * weight =  -1.0\n",
      "   - to output node [1.0, 1.0] :\n",
      "      * weight =  1.0\n",
      "\n",
      "From input node [1.0, -1.0]\n",
      "   - to output node [-1.0, 1.0] :\n",
      "      * weight =  -1.0\n",
      "   - to output node [0.0, 1.0] :\n",
      "      * weight =  -1.0\n",
      "   - to output node [1.0, 1.0] :\n",
      "      * weight =  1.0\n"
     ]
    }
   ],
   "source": [
    "for start in input_locations:\n",
    "    \n",
    "    print '\\nFrom input node', start\n",
    "    \n",
    "    for end in output_locations:\n",
    "    \n",
    "        weight = agent.evaluate(start, end)\n",
    "        print '   - to output node', end, ':'\n",
    "        print '      * weight = ', weight"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each value that is returned describes the weight of the connection from one node to another within the substrate according to the CPPN function genome that is initialized. \n",
    "\n",
    "Later on in evolution, mutations will change this function in ways that will influence the strength of these connections and fundamentally change the type of agent brain we inherit. \n",
    "\n",
    "It is customary to set some minimum weight threshold $w_t$, such that the absolute value of a weight must be greater than some value. Those weights that are less than the threshold are not expressed in the agent's final phenotype. (for each connection, if $|w| < w_t $, $ w \\gets 0 $ )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hidden nodes\n",
    "\n",
    "We have shown that a CPPN genome can map a connection to its weight by defining nodes as having coordinates within a two-dimensional substrate for a Flappy Bird-like feeding task. \n",
    "\n",
    "If we only sample these possible connections, we will generate an agent phenotype of a fully-connected artificial neural network with 6 input nodes and 3 output nodes (minus those connections that are not expressed, $|w| < w_t$ ).\n",
    "\n",
    "It may be the case that for certain tasks a neural network with no hidden nodes is sufficient to solve it, however there are plenty of tasks where layers of hidden nodes could be essential in accomplishing the task toward some objective performance measure.\n",
    "\n",
    "Since CPPNs represent continuous mappings across the entire substrate, nodes can exist at *any* location within it, not just along the top and bottom edges. \n",
    "\n",
    "There are a number of ways we can selectively sample possible hidden node locations in the substrate with respect to a given CPPN, but let's start with a simple sampling scheme that is independent of the genome itself: an evenly spaced 5 x 5 resolution grid across the substrate between input and output nodes. **Figure 5** shows what this substrate sampling looks like.\n",
    "\n",
    "## Figure 5. 5 x 5 resolution hidden node sampling\n",
    "\n",
    "![Figure 5](epann/figures/substrate1_5by5.png)\n",
    "\n",
    "Once again, our minimal architecture (inputs and outputs) is in red. Now, however, we have 25 possible hidden node locations within the substrate. We would like to determine if they would provide useful additional information for our agent genome and should be expressed in our agent phenotype network.\n",
    "\n",
    "Don't worry if our selection of these 25 hidden node locations seems arbitrary right now, because in many ways it is. The substrate is a continuous two-dimensional space, so we could have just as easily only sampled the origin (0,0) or chosen 1000 possible hidden node locations (100x100 resolution). The only difference between these two decisions of hidden node density is that if there is a lot of variation in the CPPN function (weight values vary over the possible hidden node locations from a single starting node), only sampling the origin will not describe that variation. This possible hidden node density is chosen to illustrate that a more effective hidden node sampling can be found by tying the nodes we evaluate to the function they are evaluating from - the CPPN. But we'll get there.\n",
    "\n",
    "We can once again describe the coordinate locations of these possible hidden nodes as evenly spaced within the substrate:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 364,
   "metadata": {},
   "outputs": [],
   "source": [
    "resX, resY = 5, 5\n",
    "\n",
    "x = np.linspace(-1, 1, resX)\n",
    "y = np.linspace(-1, 1, resY)\n",
    "\n",
    "xv, yv = np.meshgrid(x, y)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "And then sample the strength of the connection (evaluate weight) between the each input node and each hidden node:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 365,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "From  [-1.0, -1.0]\n",
      "[[-1. -1. -1. -1. -1.]\n",
      " [-1. -1. -1. -1. -1.]\n",
      " [-1. -1. -1.  1.  1.]\n",
      " [-1. -1.  1.  1.  1.]\n",
      " [ 1.  1.  1.  1.  1.]]\n",
      "\n",
      "From  [-0.59999999999999998, -1.0]\n",
      "[[-1. -1. -1. -1. -1.]\n",
      " [-1. -1. -1. -1. -1.]\n",
      " [-1. -1. -1. -1.  1.]\n",
      " [-1. -1.  1.  1.  1.]\n",
      " [-1.  1.  1.  1.  1.]]\n",
      "\n",
      "From  [-0.19999999999999996, -1.0]\n",
      "[[-1. -1. -1. -1. -1.]\n",
      " [-1. -1. -1. -1. -1.]\n",
      " [-1. -1. -1. -1. -1.]\n",
      " [-1. -1. -1.  1.  1.]\n",
      " [-1. -1.  1.  1.  1.]]\n",
      "\n",
      "From  [0.20000000000000018, -1.0]\n",
      "[[-1. -1. -1. -1. -1.]\n",
      " [-1. -1. -1. -1. -1.]\n",
      " [-1. -1. -1. -1. -1.]\n",
      " [-1. -1. -1. -1.  1.]\n",
      " [-1. -1.  1.  1.  1.]]\n",
      "\n",
      "From  [0.60000000000000009, -1.0]\n",
      "[[-1. -1. -1. -1. -1.]\n",
      " [-1. -1. -1. -1. -1.]\n",
      " [-1. -1. -1. -1. -1.]\n",
      " [-1. -1. -1. -1.  1.]\n",
      " [-1. -1. -1.  1.  1.]]\n",
      "\n",
      "From  [1.0, -1.0]\n",
      "[[-1. -1. -1. -1. -1.]\n",
      " [-1. -1. -1. -1. -1.]\n",
      " [-1. -1. -1. -1. -1.]\n",
      " [-1. -1. -1. -1. -1.]\n",
      " [-1. -1. -1.  1.  1.]]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAW4AAAEICAYAAAB/Dx7IAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAADM9JREFUeJzt3X+M5PVdx/Hnizu4U67ogYrVCg0W0nqRo0kTmzRUaqFC\nTaMp1jRQw9mGpDHVWGIabSHBYrU2pjGoRW2gFCExVBFiDWmo7UVJ/zDSwJkLlhQDSA9IUOS448e1\n8vaP+W4d93bZ2R+zu+/Z5yOZ3M3Md+fzyfc997y52YFNVSFJ6uOEjd6AJGl5DLckNWO4JakZwy1J\nzRhuSWrGcEtSM5s63EkqydEkn9jovUxbkoeTHEty60bvZdqc62xyrutnU4d7sLeqPrbYnUl+KcnX\nkjyfZP9SD5bksiSPDk+wO5OcutKNJflQkn9J8lKSmyc4/sNJnkxyOMlNSXbM3VdVPw783kr30tBS\nc90xnKPDwzm76pUeLMlZSb6Y5LkkTyf51Eo35lxXZc3mmuSKJPcNxz6e5FNJtq90Y7M01w7hXsp/\nAX8EfHKpA5PsAf4c+GXgdOB54DOrWPsQ8LvATROs/bPAbwFvB84EzgJ+ZxVrz7prgbMZnau3AR9J\ncvFCByY5CbgH+Arww8BrgNW8EnKu03MtE84V+F7gN4AfAH6K0Tn+zVWsPTNzbR/uqvpyVd3OaChL\nuRz4u6r6x6o6AlwDvDvJq1a49h1VdSfwnxMcfgVwY1UdrKpngOuAfStZd4u4Ariuqp6pqgeBz7L4\n+doHHKqqT1fV0ap6saoOrHRh5zpVE8+1qm6oqn+qqmNV9S3gNuAtK114lubaPtzLtAd4YO5KVT0M\nHAPOWe+1h9+fnuS0dVi7lSS7gVdz/Pnas8iXvBl4JMndw9sk+5P85LT3OXCuE1rBXOd7K3Bwrfe1\niE09160W7l3As/NuexZY0SvuVa499/v1WLubXcOv88/XYufqNcB7geuBHwH+HrhreAtl2pzr5JY7\n1+9K8n7gTcAfTmFfC9nUc20V7iR/luTIcPnoCh7iCHDKvNtOAZ6bYO27x9a+fA3Wnvv9kmvPugXm\nemS4a/75WuxcvQDcW1V3V9UxRn+4TwPeMMHaznVK1mCuc4/zC8DvA5dU1dMTrj3Tc13xd2g3QlV9\nEPjgKh7iILB37kqSs4AdwEMTrH3JKtYdX/v24fpe4KmqmuT9tpm20FyTPMHoHN0z3LSXxf+ZfIAV\nvvfpXKdnDebK8I3LzwI/V1X/uoy1Z3qurV5xLyTJtiQ7Gf0ldEKSnUlOXOTw24B3JTk/ycnAx4E7\nquq54bFunuRjQmNrbx/W3gZsG9Ze7C/DW4APJPmJJN8PXA1MvNYWdAtwdZLdSV4PXMni5+tW4M1J\nLkyyjdEnEZ4GHgTnuslMPNckP8Poz+ylVfXPC9y/dedaVZv2AhTwuiWO2TccN365eez+I8D5Y9cv\nAx4DjgJ3AaeO3fcPwJXL2N+1C6x97XDfGcPaZ4wdfxXwFHAY+BywY4HHu3Wjz/smmesORh/bOjyc\ns6vG7lvo3L4b+OZw/H5gj3PtPVfgq8B3htvmLnc71yLDBjalJC8CLwHXV9U1U17rJEbfOT63qr49\nzbUWWf8bwI8Ct1fV+9d7/fXkXGeTc13H9TdzuCVJx2v/HrckbTWGW5KaMdyS1IzhlqRmpvIf4Fx0\nwnv8jucmcc/LX8haPZZz3Tyc62yadK6+4pakZgy3JDVjuCWpGcMtSc0YbklqxnBLUjOGW5KaMdyS\n1IzhlqRmDLckNWO4JakZwy1JzRhuSWrGcEtSM4Zbkpox3JLUjOGWpGYMtyQ1Y7glqRnDLUnNGG5J\nasZwS1IzhluSmjHcktSM4ZakZgy3JDVjuCWpGcMtSc0YbklqxnBLUjOGW5KaMdyS1IzhlqRmDLck\nNWO4JakZwy1JzRhuSWrGcEtSM9s3egPdfenQ/Ru9BTXg80RryVfcktSM4ZakZgy3JDVjuCWpGcMt\nSc0YbklqxnBLUjOGW5KaMdyS1IzhlqRmDLckNWO4JakZwy1JzRhuSWrGcEtSM4Zbkpox3JLUjOGW\npGa25I8u88dIzSbnqqXMynPEV9yS1IzhlqRmDLckNWO4JakZwy1JzRhuSWrGcEtSM4Zbkpox3JLU\njOGWpGYMtyQ1Y7glqRnDLUnNGG5JasZwS1IzhluSmjHcktSM4ZakZgy3JDUzlZ85OSs/103T5fNE\nk/B5cjxfcUtSM4Zbkpox3JLUjOGWpGYMtyQ1Y7glqRnDLUnNGG5JasZwS1IzhluSmjHcktSM4Zak\nZgy3JDVjuCWpGcMtSc0YbklqxnBLUjOGW5KaMdyS1EyqaqP3IElaBl9xS1IzhluSmjHcktSM4Zak\nZlqEO0klOZrkExu9l7WU5MIkR5K8nOTCjd7PenOus8vZTleLcA/2VtXH5q4kOS/JfUmeH349b6Ev\nSrIjyY1JHk3yXJL7k1yy0k0kOSnJXyd5ZHhyXrDE8acm+dvhSfxoksvm7quqL1fVLuCxle5nBqxo\nrmPHvzfJg8P5fTjJ+SvZhHOdivmz/Ysk3xiit++VvnD4c3tTksNJnkxy1Uo3MYuz7RTu70pyEnAX\ncCuwG/g8cNdw+3zbgf8Afhr4PuBq4PYkr13FFu4F3gc8OcGxfwocA04HLgduSLJnFWvPrGXOlSQX\nAX8A/ArwKuCtwL+vYgvOdboeAH4V+PoEx14LnA2cCbwN+EiSi1ex9mzNtqo2/QUo4HVj198BfIvh\nc+jDbY8BF0/4eAeAS9dgX48DF7zC/SczegKcM3bbXwKfnHfcI8CFG32eu80V+BrwgSnsy7mu8Wzn\n3XcvsG+Jrz8EvGPs+nXAXznb0aXlK25gD3CghjM4ODDc/oqSnA6cAxyc0t7GnQN8p6oeGrvtASbY\n5xY18VyTbAPeBPxgkm8meTzJnyT5nnXYp3OdoiS7gVczOqdz1uv8tpht13DvAp6dd9uzjP65vKgk\nJwK3AZ+vqn+b0t7G7QIOz7ttyX1uYcuZ6+nAicAvAucD5wFvZPRW2LQ51+naNfw6/lxYr/PbYrZd\nw30EOGXebacAzy32BUlOYPRPnmPAhyZZJMkZw3eQjyQ5sh773OKWc75eGH7946p6oqqeBj4NvHOp\nRZzrpjc3k/FzPNH53Sqz7Rrug8C5STJ227ks8vbHcNyNjF6lXVpV355kkap6rKp2zV1WsM+HgO1J\nzh67be9i+9Tkc62qZxi9Xzn+tspE/+Md57q5DbN9gtE5nTPR+d0qs+0a7v3A/wC/PnxsaO4V9FcW\nOf4G4A3Au6rqhfl3TvIRoXnH70iyc7h6UpKd82IDQFUdBe4APp7k5CRvAX6e0St/HW8/y5vr54Bf\nS/JDw/uiHwa+OHenc91cho/l7QQCnDic38UadAtwdZLdSV4PXAncPPZYW3u2G/Vd0WV+J/i471Az\nej/zPkb/ZP468Max+z4K3D38/szh619k9M+gucvlw/0/xug9rdOWsZ9Hhsccv7x2/trD9VOBO4Gj\njD4hcdkij7flPn2wmrkO108EPgP8N6OPeV0P7HSuG39ZZLb7Fzi/Fwz3XQ4cHDt2B3DTMMOngKvG\n7tvys23xv3VN8iLwEnB9VV2zxo/9PmBPVf32Wj7uhGu/HfgbRk/Sd1bVV9d7DxvJuc4uZzvlfXQI\ntyTp/3R9j1uStizDLUnNGG5Jamb7NB70ohPe4xvnm8Q9L3/huI88rZRz3Tyc62yadK6+4pakZgy3\nJDVjuCWpGcMtSc0YbklqxnBLUjOGW5KaMdyS1IzhlqRmDLckNWO4JakZwy1JzRhuSWrGcEtSM4Zb\nkpox3JLUjOGWpGYMtyQ1Y7glqRnDLUnNGG5JasZwS1IzhluSmjHcktSM4ZakZgy3JDVjuCWpGcMt\nSc0YbklqxnBLUjOGW5KaMdyS1IzhlqRmDLckNWO4JakZwy1JzRhuSWrGcEtSM4Zbkpox3JLUjOGW\npGYMtyQ1Y7glqRnDLUnNGG5JasZwS1IzhluSmjHcktSM4ZakZgy3JDVjuCWpGcMtSc0YbklqxnBL\nUjOGW5KaMdyS1IzhlqRmDLckNbN9ozeg/+9Lh+7f6C1oCpyr1pKvuCWpGcMtSc0YbklqxnBLUjOG\nW5KaMdyS1IzhlqRmDLckNWO4JakZwy1JzRhuSWrGcEtSM4Zbkpox3JLUjOGWpGYMtyQ1Y7glqRnD\nLUnNGG5JamZL/sxJf/6fluJzZDbNylx9xS1JzRhuSWrGcEtSM4Zbkpox3JLUjOGWpGYMtyQ1Y7gl\nqRnDLUnNGG5JasZwS1IzhluSmjHcktSM4ZakZgy3JDVjuCWpGcMtSc0YbklqxnBLUjOpqo3egyRp\nGXzFLUnNGG5JasZwS1IzhluSmjHcktSM4ZakZgy3JDVjuCWpGcMtSc0YbklqxnBLUjOGW5KaMdyS\n1IzhlqRmDLckNWO4JakZwy1JzRhuSWrGcEtSM4Zbkpox3JLUjOGWpGYMtyQ1879NwNyxQlJMNQAA\nAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f82350a08d0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "count = 1\n",
    "for start in input_locations:\n",
    "    print '\\nFrom ', start \n",
    "    weights = agent.evaluate(start, [xv, yv])\n",
    "    print weights\n",
    "    \n",
    "    start_label = [ round(value, 2) for value in start]\n",
    "    \n",
    "    plt.subplot(2,3,count)\n",
    "    plt.imshow(weights)\n",
    "    plt.title(str(start_label))\n",
    "    plt.axis('off')\n",
    "    \n",
    "    count += 1\n",
    "    \n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can similarly test the weight of connections that pass from every hidden node to every output node:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 366,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "To  [-1.0, 1.0]\n",
      "[[ 1. -1. -1. -1. -1.]\n",
      " [ 1.  1.  1. -1. -1.]\n",
      " [ 1.  1.  1.  1.  1.]\n",
      " [ 1.  1.  1.  1.  1.]\n",
      " [ 1.  1.  1.  1.  1.]]\n",
      "\n",
      "To  [0.0, 1.0]\n",
      "[[ 1.  1.  1. -1. -1.]\n",
      " [ 1.  1.  1.  1.  1.]\n",
      " [ 1.  1.  1.  1.  1.]\n",
      " [ 1.  1.  1.  1.  1.]\n",
      " [ 1.  1.  1.  1.  1.]]\n",
      "\n",
      "To  [1.0, 1.0]\n",
      "[[ 1.  1.  1.  1.  1.]\n",
      " [ 1.  1.  1.  1.  1.]\n",
      " [ 1.  1.  1.  1.  1.]\n",
      " [ 1.  1.  1.  1.  1.]\n",
      " [ 1.  1.  1.  1.  1.]]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAW4AAACRCAYAAADjEVL3AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAB3dJREFUeJzt3V2IXGcdx/HvL+YNkhYSBVGQBm1KIZK0d4IUUhtrFcQL\nqRdphVxYKUW8KChYW5GKpZci+FLBEl9R1LS9Eqk2uQi9EFNsIRADhTTEGhWUpElb3/p4MSd4uu7s\nzG52Xv673w8sZc6cOc/p85z5dtg5dNNaQ5JUx4ZZn4AkaXkMtyQVY7glqRjDLUnFGG5JKsZwS1Ix\nJcKdpCW5nOSrsz6X1ZTkQJJLSd5IcmDW5zNLa2WNkzyT5PUkx2d9LvNgDa3r4SSvJTk363OBIuHu\n7GutfXHYk0k+keTZJK8mOTbqYEkOJnmpu6ieTLJznJNIsjnJz5Oc6S7K/SP235nkiW6cl5IcvPJc\na+3XrbXtwNlxxl4H3rTGSW5KcqJb0xNJbhr2wqXmeZTVvHZaax8A7h137HVi4bp+J8kfug8sh5Z6\nYZItSR5PcjHJ+ST3jztokluTHE1yIcmZMfa/Lcmp7jo4muS6K8+11g4BHx537EmrFO5R/gZ8DXh0\n1I5J9gCPAZ8E3g68CnxzGWMdB+4Gzo+x7zeAf3bj3AV8qxtfS0iyGXgK+CGwA/ge8FS3fTFXM8/T\nvHYEzwP3Ac+Nse+Xgd3AdcCtwOeT3DHmOJeBx4HPjdoxyduAI8BDwE7gd8BPxxxn+lprc/8DNOD6\nMff9FHBsxD6PAD/uPX4Pgzf9Ncs8r3PA/iWe39Yd94beth8Ajy7Y7wxwYNbzPE9rDNwO/BFIb9tZ\n4I6VzvO0rh3gEHB81nM6Dz9LvXcZfAA6NOL1LwO39x5/BfjJMs/hAHBmxD6fBp5dcE29BtzY27Yf\nODfrOW2tralP3Muxh8F/9QForb1I98Zf5XFuAP7dWjvd2/Z8N76Wtgd4oXXvmM4LLD5305znaV07\n616SHcA76M0301vXy8CLExrrqq3XcG8HLizYdgG4ZgLjXJzCOGvRctZomvM8rWtHg7mGN8+360rR\ncCf5dnc3xqUkD6zgEJeAaxdsuxZ45erPbibjrEXLmbtpzrNrOj2Xun/259t1pWi4W2v3tta2dz+P\nrOAQJ4F9Vx4keTewBTg99BUrcxrYmGR3b9u+bnwt7SSwN0l62/ay+NxNc56nde2se621vwN/ojff\nTG9dtzH4/mIu36slw72YJG9JshXYCGxIsjXJpiG7/wj4aJJbugV6GDjSWnulO9bhJIeXGGtLNxbA\n5m6sLNyv+z3ZEeDhJNuSvB/4GIMvzrS0Y8B/gM928/2ZbvszC3ccNc9JdnW3bu5abKDVvHY0WndL\n7VYgwKZuvoe16PvAg0l2JLkRuAc43DvW0Ftyk2zoxtk0eJitS9yV9ATw3iQf717zJQbfsZxayb/j\nxM3629Gr/Wa6t8+hbr/+z+He85eAW3qPDzK4S+Eyg9vOdvae+w1wzxJjnVlkrF3dcw8Av+ztuxN4\nshvnLHBwyPG8q2TBGgM3AycYfLv/HHBz77mx5xm4pZvjTZO+dnrH866S4et6bJH53t89dxdwsrfv\nFga39F0E/gzc33vuXd32tw4Ze/8i4xzrPX8SuKv3+ABwqrvejl15Ty843lzcVZLuhOZakteBfwBf\nb609NOGxNjP4dnlva+1fEx7rNuAXDC7Oj7TWjk5yvHk2yTVO8iDw19baY6t53CFjPQ28D/hta+22\nSY837ya8rncDe1prX1jN4w4Z67vAncBfWmvXT3q8kedTIdySpP9ZM7/jlqT1wnBLUjGGW5KK2TiJ\ng75xfveq/uL8Q+8c+j+F0whPv/Gz/7tNcaVWe13n2bxfc6u5rh/ccOe6Wdd5N+66+olbkoox3JJU\njOGWpGIMtyQVY7glqRjDLUnFGG5JKsZwS1IxhluSijHcklSM4ZakYgy3JBVjuCWpGMMtScUYbkkq\nxnBLUjGGW5KKMdySVIzhlqRiJvI3J1fbr17+/axPYah5/9uEWpl5vuYkP3FLUjGGW5KKMdySVIzh\nlqRiDLckFWO4JakYwy1JxRhuSSrGcEtSMYZbkoox3JJUjOGWpGIMtyQVY7glqRjDLUnFGG5JKsZw\nS1IxhluSijHcklRMib85Oc/824SSps1P3JJUjOGWpGIMtyQVY7glqRjDLUnFGG5JKsZwS1IxhluS\nijHcklSM4ZakYgy3JBVjuCWpGMMtScUYbkkqxnBLUjGGW5KKMdySVIzhlqRiDLckFWO4JakYwy1J\nxRhuSSrGcEtSMYZbkoox3JJUjOGWpGIMtyQVY7glqRjDLUnFGG5JKsZwS1IxhluSijHcklSM4Zak\nYgy3JBVjuCWpGMMtScUYbkkqxnBLUjGGW5KKMdySVIzhlqRiDLckFWO4JakYwy1JxRhuSSrGcEtS\nMYZbkoox3JJUjOGWpGIMtyQVY7glqRjDLUnFGG5JKsZwS1IxhluSijHcklSM4ZakYtJam/U5SJKW\nwU/cklSM4ZakYgy3JBVjuCWpGMMtScUYbkkqxnBLUjGGW5KKMdySVIzhlqRiDLckFWO4JakYwy1J\nxRhuSSrGcEtSMYZbkoox3JJUjOGWpGIMtyQVY7glqRjDLUnFGG5JKsZwS1Ix/wUGhWmaBgs6vAAA\nAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f823509f250>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "count = 1\n",
    "\n",
    "for end in output_locations:\n",
    "    print '\\nTo ', end \n",
    "    weights = agent.evaluate([xv, yv], end)\n",
    "    print weights\n",
    "    \n",
    "    end_label = [ round(value, 2) for value in end]\n",
    "    \n",
    "    plt.subplot(2,3,count)\n",
    "    plt.imshow(weights)\n",
    "    plt.title(str(end_label))\n",
    "    plt.axis('off')\n",
    "    \n",
    "    count += 1\n",
    "    \n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This process is recursive, and the same set of evaluations need to be executed for connections between every hidden node to every other hidden node in the given 5 x 5 substrate resolution.\n",
    "\n",
    "### Break - substrate wrt CPPNs\n",
    "\n",
    "Let's take a moment to set aside the idea that we are arbitrarily designating a resolution by which we are sampling hidden node locations (and therefore possible connections) and look at all possible connection evaluations from each of the input nodes (at a higher resolution of 1000 x 1000 -> 1,000,000 sampled hidden nodes)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 367,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYkAAAEICAYAAACqMQjAAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAD8xJREFUeJzt3HusZfVZxvHnmbsyRQfUsVqhoYW0TmRo0sQmDZVaqFDT\naIo1DVPD2IakMdXYiWm0hWQsVmtjGoNa1AZKERJDlUKsmTTUdqKkfxhpYMwES4oBpAMkKDIXLtOW\nt3/sdejumt87Z+37unw/yc7M2XufvVbW+5zfw2+fPTgiBABAyYZVnwAAoL0oCQBAipIAAKQoCQBA\nipIAAKQoCQBAajAlYTtsn7D98VWfy6LZftj2Sdu3rfpcFo259hNzbY/BlERld0R8NHvQ9q/b/prt\n52wfXO/FbF9l+9EqzHfZPmvaE7P9Qdv/YftF27c0eP6HbD9p+6jtm21vXXssIl4j6Y+nPZcOWm+u\nW6trdLS6ZvtO92K2z7P9RdvHbD9t+5PTnhhzncnc5mr7atv3Vc993PYnbW+a9sSGNNehlcR6/k/S\nn0v6xHpPtL1L0t9I+g1JOyU9J+nTMxz7iKQ/knRzg2P/kqTfl/Q2SedKOk/SH85w7L7bL+l8ja7V\nWyV92PblpSfa3iLpHklfkfSTkl4laZb/wmOui7NfDecq6Ycl/a6kH5P08xpd49+b4diDmSslMSYi\nvhwRd2gUgPXskfRPEfGvEXFc0nWS3mX7FVMe+86IuEvS/zZ4+tWSboqIwxHxjKTrJe2d5rgDcbWk\n6yPimYh4UNJnlF+vvZKORMSnIuJERLwQEYemPTBzXajGc42IGyPi3yLiZER8S9Ltkt487YGHNFdK\nYnq7JD2w9kVEPCzppKQLln3s6u87bZ+9hGN3iu0dkl6pU6/XruRb3iTpEdsHqreaDtr+uUWfZ4W5\nNjTFXOveIunwvM8r0em5UhLT2y7p2dp9z0qaaicx47HX/r6MY3fN9urP+vXKrtWrJL1H0g2SfkrS\nP0u6u3obatGYa3OTzvVltt8n6Y2S/mwB51XS6bkOtiRs/7Xt49XtI1O8xHFJZ9buO1PSsQbHPjB2\n7D1zOPba39c9dt8V5nq8eqh+vbJr9bykeyPiQESc1GghOVvS6xscm7kuyBzmuvY6vyrpTyRdERFP\nNzz2oOc69W/3uy4iPiDpAzO8xGFJu9e+sH2epK2SHmpw7CtmOO74se+ovt4t6amIaPL+aK+V5mr7\nCY2u0T3VXbuVv9VwSFO+V81cF2cOc1X1S+3PSPrliPjPCY496LkOdidRYnuj7W0alecG29tsb06e\nfrukd9q+2PYZkj4m6c6IOFa91i1NPho3duxN1bE3StpYHTsr8Vslvd/2z9r+UUnXSmp8rAG6VdK1\ntnfYfp2ka5Rfr9skvcn2pbY3avSJmKclPSgx15ZpPFfbv6jRz+yVEfHvhceZayYiBnGTFJJeu85z\n9lbPG7/dMvb4cUkXj319laTHJJ2QdLeks8Ye+xdJ10xwfvsLx95fPXZOdexzxp6/T9JTko5K+qyk\nrYXXu23V170lc92q0UcVj1bXbN/YY6Vr+y5J36yef1DSLuba7blK+qqk71T3rd0OMNf1b65OsPds\nvyDpRUk3RMR1Cz7WFo0+wXBhRHx7kcdKjv8NST8t6Y6IeN+yj79MzLWfmGt7DKYkAACT43cSAIAU\nJQEASFESAIAUJQEASLX6H9NdtuHd/Fa9Je556fOe12sx1/aY51wlZtsm85otOwkAQIqSAACkKAkA\nQIqSAACkKAkAQIqSAACkKAkAQIqSAACkKAkAQIqSAACkKAkAQIqSAACkKAkAQIqSAACkKAkAQIqS\nAACkKAkAQIqSAACkKAkAQIqSAACkKAkAQIqSAACkKAkAQIqSAACkKAkAQIqSAACkKAkAQIqSAACk\nNq36BNBuXzpy/6pPAcAKsZNAioIAQEmgiIIAIPF2E2ooBwDj2EngZRQEgDp2EqAcAKTYSQwcBQHg\ndNhJDBTlAKAJdhIDREEAaIqdxIBQDgAmxU5iICgI1JEJNMFOoudYCFBHJjAJdhI9xmKAOjKBSbGT\n6CEWAtSRCUyLnUTPsBigjkxgFuwkeoKFAHVkAvPATqIHWAxQRyYwL+wkOoyFACXkAvNESXQUCwHq\nyAQWgZLoGBYClJALLAol0SEsBKgjEyiZZy4oiQ5gIUAJuUDJvHNBSbQcCwHqyARKFpULPgLbYiwG\nqCMTKFlkLthJtBALAerIBEqWkQt2Ei3DYoA6MoGSZeWCnURLsBCgjkygZNm5YCfRAiwGqCMTKFlF\nLthJrBALAerIBEpWmQt2EivCYoA6MoGSVeeCncSSrXrgaB8ygZK25IKdxBK1ZehoDzKBkjblgp3E\nErRp4GgHMoGSNuaCncSCtXHoWC0ygZK25oKdxIK0deBYHTKBTJuzQUksQJsHjtUgEyjpQi4oiTnq\nwsCxXGQCma5kg5KYk64MHMtDJlDStVxQEjPq2sCxHOQCJV3MBSUxgy4OHItFJlDS5VxQElPo8sCx\nOOQCJV3PBSUxoa4PHPNHJlDSl1zwj+km0JehY37IBEr6lAt2Eg30aeCYDzKBkj7mgp3EOvo4dMyG\nTKCkr7lgJ5Ho68AxPTKBkr7ngp1EQd+HjsmRCZQMIRfsJMYMYeCYDJlAyZBywU6iMqShoxkygZKh\n5WLwO4mhDRzrIxMoGWouBr2TGOrQkSMTKBlyLga5kxjywFFGJlBCLga4k2DoqCMTKCEXI4PZSTBw\n1JEJlJCLHzSInQRDRx2ZQAm5OFWvdxIMHHVkAhmyUdbbkmDgqCMTKCEXp9e7kmDgKCEXKCEX6+tV\nSTBw1JEJlJCL5npREgwcdWQCGbIxmc6XBANHHZlACbmYTmdLgoGjhFyghFxMr5MlwcBRRyZQQi5m\n17l/TMfQUUcmUEIu5qMzOwkGjjoygRJyMV+d2EkwdNSRCZSQi/lr9U6CgaOOTKCEXCxOJ3YSgMRC\ngDJysVit3kkAEosAysjFcjgiVn0OAICW4u0mAECKkgAApCgJAECKkgAApAZXErbD9gnbH1/1ucyT\n7UttH7f9ku1LV30+y8Zc+4m5rt7gSqKyOyI+uvaF7Yts32f7uerPi0rfZHur7ZtsP2r7mO37bV8x\n7UnY3mL7H2w/Uv0wXLLO88+y/YXqh+ZR21etPRYRX46I7ZIem/Z8emCquY49/z22H6yu78O2L57m\nJJjr3NXn+re2v1EtsHtP943Vz+zNto/aftL2vmlPYqhzHWpJvMz2Fkl3S7pN0g5Jn5N0d3V/3SZJ\n/yPpFyT9iKRrJd1h+9UznMK9kt4r6ckGz/0rSScl7ZS0R9KNtnfNcOzemnCusn2ZpD+V9JuSXiHp\nLZL+e4ZTYK6L84Ck35L09QbP3S/pfEnnSnqrpA/bvnyGYw9vrhExqJukkPTasa/fLulbqv7NSHXf\nY5Iub/h6hyRdOYfzelzSJad5/AyNAnfB2H1/J+kTtec9IunSVV/nrs1V0tckvX8B58Vc5zjX2mP3\nStq7zvcfkfT2sa+vl/T3zLX5bfA7CUm7JB2KamKVQ9X9p2V7p6QLJB1e0LmNu0DSdyLiobH7HlCD\n8xyoxnO1vVHSGyX9uO1v2n7c9l/a/qElnCdzXRDbOyS9UqPruWZZ17Y3c6UkpO2Snq3d96xGbzmk\nbG+WdLukz0XEfy3o3MZtl3S0dt+65zlgk8x1p6TNkn5N0sWSLpL0Bo3eTlw05ro426s/x3OwrGvb\nm7lSEtJxSWfW7jtT0rHsG2xv0GjreFLSB5scxPY51acZjts+vozzHLhJrtfz1Z9/ERFPRMTTkj4l\n6R3rHYS5ttraPMavb6Nry1y/j5IYvVV0oW2P3XehkreQqufdpNF/fV4ZEd9ucpCIeCwitq/dpjjP\nhyRtsn3+2H27s/NE87lGxDMavcc8/tZUo/+pGXNtr2quT2h0Pdc0urbM9fsoCemgpO9K+p3q43Jr\nO4OvJM+/UdLrJb0zIp6vP9jko3G152+1va36covtbbWFTZIUESck3SnpY7bPsP1mSb+i0Y4Gpzqo\nyeb6WUm/bfsnqveyPyTpi2sPMtf2qD6Kuk2SJW2urm22lt0q6VrbO2y/TtI1km4Zey3mup5V/+Z8\n2TcVPi2h0fvP92n0tsPXJb1h7LGPSDpQ/f3c6vtf0Gg7uXbbUz3+Mxq9D3n2BOfzSPWa47dX149d\nfX2WpLskndDokzpXJa/X6k9LtG2u1debJX1a0v9r9PHGGyRtY66tnOvBwrW9pHpsj6TDY8/dKunm\nan5PSdo39hhzbXAb3P8q3PYLkl6UdENEXDfn136vpF0R8QfzfN2Gx36bpH/U6IfiHRHx1WWfwyox\n135irqs3uJIAADTH7yQAAClKAgCQoiQAAKlNqz6B07lsw7v5hUlL3PPS50/5mN+0mGt7zHOuErNt\nk3nNlp0EACBFSQAAUpQEACBFSQAAUpQEACBFSQAAUpQEACBFSQAAUpQEACBFSQAAUpQEACBFSQAA\nUpQEACBFSQAAUpQEACBFSQAAUpQEACBFSQAAUpQEACBFSQAAUpQEACBFSQAAUpQEACBFSQAAUpQE\nACBFSQAAUpQEACBFSQAAUpQEACBFSQAAUpQEACBFSQAAUpQEACBFSQAAUpQEACBFSQAAUpQEACBF\nSQAAUpQEACBFSQAAUpQEACBFSQAAUpQEACBFSQAAUpQEACBFSQAAUpQEACBFSQAAUpQEACBFSQAA\nUpQEACBFSQAAUpQEACBFSQAAUpQEACBFSQAAUpQEACBFSQAAUpQEACBFSQAAUptWfQJoty8duX/V\npwBghdhJIEVBAGAngVNQDgDWsJPAD6AgAIxjJwFJlAOAMnYSoCAApNhJDBjlAGA97CQGioIA0AQ7\niYGhHABMgp3EgFAQACbFTmIAKAfUkQk0xU6i51gMUEcmMAl2Ej3FQoA6MoFpsJPoIRYD1JEJTIud\nRI+wEKCEXGAWlERPsBCgjkxgHni7qQdYDFBHJjAv7CQ6jIUAdWQC80ZJdBALAUrIBRaBkugYFgLU\nkQksEr+T6BAWA9SRCZTMMxfsJDqAhQB1ZAIli8gFO4mWYzFAHZlAyaJywU6ipVgIUEcmULLoXLCT\naCEWA9SRCZQsIxfsJFqEhQB1ZAIly8wFO4mWYDFAHZlAybJzwU5ixVgIUEcmkFlFNiiJFWIxQB2Z\nQMkqc0FJrAALAerIBErakAt+J7FkbRg62oVMoKQtuWAnsSRtGTjag0wg06ZsUBJL0KaBox3IBEra\nmAtKYoHaOHCsHrlASVtzQUksSFsHjtUhEyhpey4oiTlr+8CxGuQCJV3IBSUxR10YOJaLTKCkS7ng\nI7Bz0qWhYznIBEq6lgtHxKrPAQDQUuwkAAApSgIAkKIkAAApSgIAkKIkAAApSgIAkKIkAAApSgIA\nkKIkAAApSgIAkKIkAAApSgIAkKIkAAApSgIAkKIkAAApSgIAkKIkAAApSgIAkKIkAAApSgIAkKIk\nAAApSgIAkKIkAACp7wG3ViYi6mbVGQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f82353cd490>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "resX, resY = 1000, 1000\n",
    "\n",
    "x = np.linspace(-1, 1, resX)\n",
    "y = np.linspace(-1, 1, resY)\n",
    "\n",
    "xv, yv = np.meshgrid(x, y)\n",
    "\n",
    "for node in range(len(input_locations)):\n",
    "\n",
    "    start = input_locations[node]\n",
    "    weights = agent.evaluate(start, [xv, yv])\n",
    "    \n",
    "    start_label = [ round(value, 2) for value in start]\n",
    "\n",
    "    plt.subplot(2,3, (node+1))\n",
    "    plt.imshow(weights)\n",
    "    plt.title(str(start_label))\n",
    "    plt.axis('off')\n",
    "    \n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As well as the substrate evaluation values for all possible connections coming into the ouput nodes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 368,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYkAAACRCAYAAAA2L0T/AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAACBBJREFUeJzt3FuopXUdxvHvMzkH8ABOQRSEUirChIe7IIQxJ7Mgugi7\nGA3mIiMiuhAKMo0wEi8j6GCQTEeKysNViKVzIV5EigoDJgijmFlBoc6onfx3sd4d7+z2b6+196zT\nftf3A4vZ633f9f7e+R/2s94DO601JEnayK5FH4AkaXkZEpKkkiEhSSoZEpKkkiEhSSoZEpKk0sqF\nRJKW5FSSry/6WKYpyaEkJ5O8meTQoo9nUYbSv0keSvJGkkcWfSzLYED9ejTJ60leWPSxTGrlQqJz\neWvty9XKJJ9I8miS15IcG7ezJIeTPNcN4vuS7J/kIJLsSfLLJCe6SXBwzPb7k9zb1XkuyeG1da21\n37TWzgGen6T2wJ3Wv0muSPJY15+PJbmi+uBmbTzONMdNa+0DwGcmrb0i1vfr95L8oftidGSzDybZ\nm+TuJK8keSnJzZMWTXJ1koeTvJzkxATbX5Pk6W4cPJzkgrV1rbUjwIcnrb0MVjUkxvkb8A3gznEb\nJjkA3AV8Eng78Brw7S3UegS4EXhpgm2/Bfyzq3MD8J2uvgpJ9gD3Az8Gzgd+ANzfLd/ImbTxPMeN\n4Engs8DjE2z7VeBi4ALgauCLSa6bsM4p4G7gC+M2TPI24B7gNmA/8Hvg5xPWWU6ttZV6AQ24aMJt\nPwUcG7PNHcBPe+/fw+iXzLlbPK4XgIObrD+72+8lvWU/Au5ct90J4NCi23lZ+he4FvgjkN6y54Hr\nttvG8xo3wBHgkUW36TK8Npu3jL5oHRnz+ReBa3vvvwb8bIvHcAg4MWabTwOPrhtTrwOX9pYdBF5Y\ndJtO+vJM4swdYPSNBoDW2rN0v2imXOcS4N+ttWd6y57s6qt2AHiqdbOz8xQbt9s823he42blJTkf\neAe99mZ+/XoKeHZGtebCkDhz5wAvr1v2MnDuDOq8Moc6Q7OV/plnG89r3GjU1nB6e9uvE1r5kEjy\n3e6poJNJbtnGLk4C561bdh7w6pkf3ULqDM1W2m2ebWx/zs/J7t9+e9uvE1r5kGitfaa1dk73umMb\nuzgOXL72Jsm7gb3AM+UntucZ4KwkF/eWXd7VV+04cFmS9JZdxsbtNs82nte4WXmttb8Df6LX3syv\nX89mdL9px87TlQ+JjSR5S5J9wFnAriT7kuwuNv8J8NEkV3UD4nbgntbaq92+jiY5ukmtvV0tgD1d\nrazfrru2eQ9we5Kzk7wf+BijG6uqHQP+A3y+a+vPdcsfWr/huDZOcmH3qPKFGxWa5rjReN0j5PuA\nALu79q5+p/0QuDXJ+UkuBW4Cjvb2VT6CnmRXV2f36G32bfJ03L3Ae5N8vPvMVxjdE3t6O//HpbDo\nO+fzfjHB002Mnipp615He+tPAlf13h9m9MTMKUaPW+7vrfstcNMmtU5sUOvCbt0twK972+4H7uvq\nPA8cLvbn002nL7sSeIzRUyaPA1f21k3cxsBVXfvunvW46e3Pp5vqfj22QXsf7NbdABzvbbuX0WOs\nrwB/Bm7urXtXt/ytRe2DG9Q51lt/HLih9/4Q8HQ33o6tzed1+9sxTzelO+iVkeQN4B/AN1trt824\n1h5GTzpc1lr714xrXQP8itFk+Ehr7eFZ1ltWs+zfJLcCf22t3TXN/Ra1HgTeB/yutXbNrOstuxn3\n643Agdbal6a536LW94Hrgb+01i6adb1pWLmQkCRNznsSkqSSISFJKhkSkqTSWYs+gM28+dLFG94w\n+dA7yz/iqRl58M1f/N9judtV9avmP7an2a8AH9x1vX27JKbVt0sdEpUHXnzifz8bGBqStbHtuNay\n2JEh0eek0hD5RUjLYseHxBonlYbKsa1FGkxI9DmpNFSObc3bIEOiz0mlofJSq+Zh8CHR56TSEPlF\nSLO0UiGxxkmloXJsa9pWMiT6nFQaKs+cNQ0rHxJ9TioNkV+EdCYMiQ04qTRUjm1tlSExhpNKQ9Uf\n21LFkNgCA0PSqvGvwG7TAy8+4TcxSYPnmcQZ8uxC0pAZElNkYEgaGi83zYiXoyQNgWcSM+bZhaSd\nzJCYIwND0k5jSCyIgSFpJzAkloCBIWlZeeN6yXjDW9Iy8UxiSXl2IWkZGBI7gIEhaVG83LTDeDlK\n0jx5JrFDeXYhaR4MiQEwMCTNipebBsbLUZKmyTOJgfLsQtI0GBIrwMCQtF2GxIoxMCRthSGxwgwM\nSeN441qAN7wlbcwzCZ3GswtJfYaESp5ZSPJykySpZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSp\nZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSpZEhI\nkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkqG\nhCSpZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSp\nZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSpZEhI\nkkqGhCSpZEhIkkqGhCSpZEhIkkqGhCSpZEhIkkpprS36GCRJS8ozCUlSyZCQJJUMCUlSyZCQJJUM\nCUlSyZCQJJUMCUlSyZCQJJUMCUlSyZCQJJUMCUlSyZCQJJUMCUlSyZCQJJUMCUlSyZCQJJUMCUlS\nyZCQJJUMCUlSyZCQJJUMCUlSyZCQJJUMCUlS6b+DHpRvgs6CaQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f8237f24390>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for node in range(len(output_locations)):\n",
    "\n",
    "    end = output_locations[node]\n",
    "    weights = agent.evaluate([xv, yv], end)\n",
    "    \n",
    "    end_label = [ round(value, 2) for value in end]\n",
    "\n",
    "    plt.subplot(1,3, (node+1))\n",
    "    plt.imshow(weights)\n",
    "    plt.title(str(end_label))\n",
    "    plt.axis('off')\n",
    "    \n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And for the fun of it, let's look at what connections that start and some random hidden node locations look like."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 369,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAVQAAAEICAYAAAAA3gw5AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAEfdJREFUeJzt3HuQZGV5x/HvDxaWyLIKqCjgSqEQDYlAQiJK1qCAopFE\nQ0wZNgjxEo21iYrRRISq1SBEov6BUfFSiAgaY0WlMIV4RZREAS3BICXxggQRBMVdFrlEefNHn056\nm+6Znpl3pvt0fz9VXbtz3tOnn9l9+jfPOadrUkpBkrR02427AEmaFgaqJFVioEpSJQaqJFVioEpS\nJQaqJFUyNYGapCS5K8mbxl3LJEvy+ST3JPnyuGvRwtjjD5Rk/yRbk/wyyYvHXc/UBGrjwFLK64ct\nJlmd5JwkW5LckuSkOfY9u/mP6j7uTXJnz/qlTTB117+9lMKTHJfkB80b5hNJdptj3/ck+XaS+5Oc\n2Ld2QpKvNd/jTUnOTLKqu15KeRrwsqXUqrGq1uPN/vsm+WSSO5PcnuTMvvXnJ7mu6cvvJlk/SpHp\neHOSnzSPNyfJHPsP7f8kG5Nc1bwHz+19Xinl+lLKGuBLo9S13KYtUOezCdgPeDTwVOC1SY4etGMp\n5WWllDXdB/Bh4KN9u23s2edXF1tUkgOAdwPHA3sAPwfeOcdTrgZeDnx9wNqDgFcCDwWeCBwB/M1i\na1PrbGLEHk+yI/AZ4PPAI4C9gfN71o8C3gz8ObAL8BTgeyPW8RfAc4ADgScAxwAvHVLHfP1/M3Aa\ncM6Irz02q+bfZaqcAJxYSrkDuCPJe4ETgU/N9aQkOwPHAs9epro2ABeVUi5rXu9U4Loku5RS7uzf\nuZTyjma/ewasvavnyx8muYDOG0uzYSE9fiJwcynlbT3brun5+xuAN5ZSvtJ8/cMF1vHWUspNAEne\nCrwEOHvAvnP2fynlY832Q+iE/sSamQk1ya7AI+lMd11XAweM8PRjgduAy/q2n9GcJl2e5PAllHdA\nb12llO8C9wH7L+GYXU8Brq1wHE24RfT4ocANSS5u+vjSJL/RHGt74BDgYUm+01w++qckvzJiOdv0\n9Dx1LGf/r6iZCVRgTfPn5p5tm+mcysznBOC8su0vPvhbYF9gL+A9wEVJHrOE2jb3bRu1tqGSvJDO\nm+ItSzmOWmOhPb438HzgLGBP4N+AC5tLAXsAOwB/DKwHDgIOBk5ZQC39dawZch11Wfp/HKY2UPtu\nKp0MbG2W1vbsthZ4wCl133HWAYcD5/VuL6V8tTkdubeU8gHgcuBZI9S1vqeu7uS4ta+ukWqb53We\nA5wBPLOUcvtij6PJVaHH7wa+XEq5uJRyH50fvLsDj2/WAN5eSvlR00NvY4Qeb/T39Fpga99QMmzf\n+eqeWFMbqH03lU5vrin9iM5F8q4Dmf90+Hjg8lLKfBfjCzD0LmZPXV/qqat7CnRtb11J9gVWA9fP\nd7xBmpsQ7wWOKaV8czHH0OSr0OPX0OnbQce+A7ipb30hv5pum56ep46q/T9OUxuoQ5wHnJJk1ySP\no3OR/Nx5nvOC/n2SPCTJM5LslGRVkg10rlV+qlnfJ53PDO4zYl0XAMc00+vOwBuBjw26IdUcf8ck\nO9EJ8B2aOrZr1p7WHO/YUsoVI76+psdCevx84NAkRzbXTF8J3A5c16y/H/irJA9vrs++Cvhk98lN\njx8+Rx0nJdkryZ7Aq+eoY87+b95jOwHbA9t333fz/kuMQyllKh50fno+dp59VtP56MUW4FbgpJ61\ndXROPdb1bHsScBewS99xHgZcSeeU5GfAV4CjetbXAzcAOyyg/uOAG5vXuxDYrWftYuDknq8vbb7f\n3sfhzdoXgF8030v3cXHfa51I51Rv7P9vPkZ/LFOP/xHwnWb/S4EDetZ2oPPxpZ8Bt9C51rpTs/ao\n5jm7D6kjwJnAT5vHmUB61rcC63u+nqv/Nw3o9019r3cp8OJx/x+lKab1mo8Q3QucVUo5dcy1nALc\nVkp59zjrGCTJZ+jc3b2ilHLEuOvR6Casx/+MTvi+bsx17EdnuNkReHkp5dyx1jMtgSpJ4zZr11Al\nadkYqJJUiYEqSZVM5kcPGkdt97yZusB7yc3fGHcJQ233iP+a9zO2Gt209PYk9+yoava2E+qEmIbG\n1GyxZx9ooifUWWFjqk3s1+EM1DGyMdU29uzcDNQxsTHVJvbraAzUFWZjqm3s2dEZqCvIxlSb2K8L\nZ6CuABtTbWPPLo6BusxsTLWJ/bo0BuoysTHVNvbs0hmoy8DGVJvYr/UYqBXZmGobe7YuA7USG1Nt\nYr8uDwN1iWxMtY09u3wM1CWwMdUm9uvyM1AXwcZU29izK8Nf37dANqbaxp5dOU6oC2Bjqk3s15Vn\noI7AxlTb2LPjYaDOw8ZUm9iv42WgDmFjqk3s18ngTakBbE61if06OZxQe9iYaht7drIYqA0bU21i\nv06mmQ9UG1NtY89OrpkOVBtTbWK/Tr6ZDFQbU21jz7bDzAWqjak2sV/bZWYC1cZU29iz7TMTgWpj\nqk3s1/aa6kC1MdU29my7TW2g2phqE/t1OkxdoNqYaht7dnpMVaDamGoT+3X6TEWg2phqG3t2OrX+\nt03ZmJImRasD1TCVNElaecpvkEqaRK2bUA1TSZOqNROqQSpp0k18oBqkktpiok/5DVNJbTLRgSpJ\nbWKgSlIlBqokVWKgSlIlBqokVWKgSlIlBqokVWKgSlIlBqokVWKgSlIlBqokVWKgSlIlBqokVWKg\nSlIlBqokVWKgSlIlBqokVWKgSlIlBqokVWKgSlIlBqokVWKgSlIlBqokVWKgSlIlBqokVWKgSlIl\nBqokVWKgSlIlBqokVWKgSlIlBqokVWKgSlIlBqokVWKgSlIlBqokVWKgSlIlBqokVWKgSlIlBqok\nVWKgSlIlBqokVWKgSlIlBqokVWKgSlIlBqokVWKgSlIlBqokVWKgSlIlBqokVWKgSlIlBqokVWKg\nSlIlBqokVWKgSlIlBqokVWKgSlIlBqokVZJSyrhrkKSp4IQqSZUYqJJUiYEqSZUYqJJUyVQGapKS\n5K4kbxp3LZMiyZFJtia5P8mR465HC2dfD5bku0nuS3L+uGuZykBtHFhKef2wxSSrk5yTZEuSW5Kc\nNMpBk3yuaexVPdtuSHJ3E1hbk3x61CKT7Jbk480b5QdJjptn/99MclnzOrcmeUXP2pOTXJHkziTX\nJPnd7lop5bOllDXAjaPWpom0TV8nOSjJ15L8vPnzoGFPTLIxyVVJ7k1y7oD1P0lyXdM/30rynKUU\nmuRVzXtrS/NeWz3Hvg9K8s4ktyfZnOSynrXXJPnPpq7vJ3lN73NLKY8BTl9KrbVMc6DOZxOwH/Bo\n4KnAa5McPdcTkmwAdhiyfEwpZU3zePoC6ngHcB+wB7ABeFeSA4a8/kOBTwHvBnYHHgt8ulnbDbgI\n+EfgIcCZwEVJdl1ALWqRJDsCFwLnA7sCHwAubLYPcjNwGnDOgGPt1RznJGAt8BrgQ0kevsjangH8\nHXAEnffYvsAb5njKe4DdgMc3f76q93DAC+h8j0cDG5M8fzF1LbtSytQ9gAI8dp59bgae3vP13wP/\nPMf+DwauBw5tjr+qZ+0G4MhF1LkznTDdv2fbB4F/GLL/6cAHh6w9G7i2b9v1wIv6ti2qVh/jf/T3\nNfB04Ic0nydvtt0IHD3PcU4Dzu3b9kTgx33bbgOetMhaPwSc3vP1EcAtQ/Z9HLAFWDvisc8C3t63\nbRNw/rj/j2ZyQm2mtkcCV/dsvhoYOBk2TgfeBdwyZP2CJLcl+XSSA0csZX/gF6WU60es41Dgp0n+\nPcmPk1yUZF3Pevr2D/DrI9ai9jkAuKY0idK4hrn7eJirgOuS/EGS7ZvT/Xub4y22tv731x5Jdh+w\n7+8APwDe0JzyfzPJsYMOmiTAeuDaRda1rGYyUIE1zZ+be7ZtBnYZtHOSQ4DDgLcPOd4GYB86pzZf\nAC5J8pAR69jSt21oHcDewAnAK4B1wPeBDzdr/wHsmeRPk+yQ5ATgMcCDRqhD7bSGbXsY5u6foUop\nvwTOozNZ3tv8+dJSyl2Vauv+fVBte9P5wb8Z2BPYCHwgyeMH7LuJTm69f5F1LauZCNQkZ/fcMDoZ\n2Nosre3ZbS1w54Dnbge8E3hFKeUXg45fSrm8lHJ3KeXnpZQzgJ/R+Sk6n619NQyto3E38PFSypWl\nlHvoXJN6cpIHl1J+AvwhnWtgt9K51vRZ4KYR6lA7LbR/hmo++XEmcDiwI/B7wPvmusnV89wNPe+v\ni4fU1v37oNruBv4HOK2Ucl8p5Yt0BpNt7kUk2UjnWurvl1LuHfmbW0EzEaillJeV/79hdHop5Q7g\nR0DvqfmBDD6NWAscAnwkyS3Alc32m5IMC83CA0+/B7keWJVkvxHqgM7pV+/p3Ta/iKGU8sVSym+X\nUnYDjqdzbeqKEepQO10LPKE5De56Aos7HT4IuKyUclUp5f5SypXAV4F5P2JXSrmg5/31zJ7a+t9f\ntzY/+PsNuqywTW8neSHNTa5SysQOCTMRqEOcB5ySZNckjwNeApw7YL/uachBzeNZzfbfAr6aZF2S\nw5LsmGSn5iMdDwUuB0hyeJKBv4GmOZ36GPDGJDsnOYzOlPnBITW/H3hu81GZHYBTgS+XUjY3r3Vw\nc7q/FngL8N+llEsW9K+iNrkU+CXw1+l8DHBjs/3zg3ZOsirJTsD2wPZNv3Y//nclsL47kSY5mM5Z\n1jXN10P7eIjzgBcl+bXm8tcpDH5/AVxG52ba65oaD6PzyZtLmtfeQOcexlGllO8toIaVN+67Ysvx\nYLS7/KvpfHxkC51T5JN61tbROWVZN+B5+9Bzl5/mxgBwF/AT4HPAIT37Hw9cPkcduwGfaJ5/I3Bc\nz9p6YGvf/n9J587uHXQ+JvWonrUP0/kBsBn4CPDwAa93A97lb+VjUF8DBwNfo3Pa/HXg4J61k4GL\ne77e1Byj97GpZ30j8B06p+XfA17dszZnHw+pt3v5aQudYWB1z9q1wIaerw+gcx/gLuBbwHN71r5P\n55LA1p7H2X2vtYkJuMs/lb++L8k9dC6sn1VKOXXMtbwP+GgZ86SY5AjgX+n8IHlWKeUL46xHCzfO\nvp6UPh4kybeBvYB/KaW8cKy1TGOgStI4zPI1VEmqykCVpEoMVEmqZNX8u4zPUds9zwu8E+Iz9390\nlM/VakT29uSo2dtOqJJUiYEqSZUYqJJUiYEqSZUYqJJUiYEqSZUYqJJUiYEqSZUYqJJUiYEqSZUY\nqJJUiYEqSZUYqJJUiYEqSZUYqJJUiYEqSZUYqJJUiYEqSZUYqJJUiYEqSZUYqJJUiYEqSZUYqJJU\niYEqSZUYqJJUiYEqSZUYqJJUiYEqSZUYqJJUiYEqSZUYqJJUiYEqSZUYqJJUyapxFyBJ43DJzd+o\nfkwnVEkzZznCFJxQJc2Q5QrSLidUSTNhucMUDFRJM2AlwhQ85Zc0xVYqSLucUCVNpZUOU3BClTRl\nxhGkXU6okqbGOMMUnFAlTYFxB2mXE6qkVpuUMAUnVEktNUlB2uWEKql1JjFMwQlVUotMapB2OaFK\naoVJD1NwQpU04doQpF1OqJImVpvCFJxQJU2gtgVplxOqpInS1jAFJ1RJE6LNQdrlhCpp7KYhTMEJ\nVdIYTUuQdjmhShqLaQtTcEKVtMKmMUi7nFAlrZhpDlNwQpW0AqY9SLucUCUtq1kJUzBQJS2jWQpT\n8JRf0jKYtSDtckKVVNWshik4oUqqZJaDtMsJVdKSGaYdTqiSFs0g3ZYTqqRFMUwfyAlV0oIYpMM5\noUoamWE6NydUSfMySEfjhCppTobp6JxQJQ1kkC6cE6qkBzBMF8cJVdL/MUiXxglVEmCY1uCEKs04\ng7QeJ1RphhmmdRmo0owyTOvzlF+aMQbp8nFClWaIYbq8nFClGWCQrgwnVGnKGaYrxwlVmlIG6cpz\nQpWmkGE6Hk6o0hQxSMfLCVWaEobp+DmhSi1nkE4OJ1SpxQzTyeKEKrWQQTqZnFClljFMJ5cTqtQS\nBunkc0KVWsAwbQcnVGmCGaTt4oQqTSjDtH2cUKUJY5C2lxOqNEEM03ZzQpUmgEE6HZxQpTEzTKdH\nSinjrkGSpoITqiRVYqBKUiUGqiRVYqBKUiUGqiRVYqBKUiUGqiRVYqBKUiUGqiRVYqBKUiUGqiRV\nYqBKUiUGqiRVYqBKUiUGqiRVYqBKUiUGqiRVYqBKUiUGqiRVYqBKUiUGqiRVYqBKUiUGqiRV8r/z\nu5piCNll9gAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f8237c0c190>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_random_hiddens = 4\n",
    "\n",
    "hidden_locations = [ np.random.uniform(-1, 1, size=[1,2]) for location in range(num_random_hiddens)]\n",
    "\n",
    "for node in range(len(hidden_locations)):\n",
    "\n",
    "    start = hidden_locations[node][0]\n",
    "    start = [ start[0], start[1] ]\n",
    "    \n",
    "    weights = agent.evaluate(start, [xv, yv])\n",
    "        \n",
    "    start_label = [ round(value, 2) for value in start]\n",
    "\n",
    "    plt.subplot(2,2, (node+1))\n",
    "    plt.imshow(weights)\n",
    "    plt.title(str(start_label))\n",
    "    plt.axis('off')\n",
    "    \n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Each agent is initialized with a simple neural network that represents its genome. Hyper-NEAT, the algorithm we will be using to evolve these neural networks, is unique in it's ability to add complexity to neural networks by probabilistically adding nodes and connections, allowing for the possibility for the functions they calculate to become incrementally more complex. It is often the case in other neuroevolution approaches to initialize a population with randomly varying network architectures at the outset, though the justification for doing so is a little unclear. CPPNs in a population here are initialized as neural networks with 5 input nodes and 1 output node (no hidden nodes), that way agents in the initial population only differ in the weights of their connections."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
